File: 101~200/M-102-BinaryTree LevelOrderTraversal/JeongYe.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public Map<Integer, ArrayList> levels = new HashMap<>(); // 트리 레벨, 해당 레벨의 노드 값 목록 저장
    public int depth = 0; //트리 레벨

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> answer = new ArrayList<>();

        if(root == null) return new ArrayList<>();

        traverse(root);

        for(int depth : levels.keySet()) {
            answer.add(levels.get(depth));
        }

        return answer;

    }

    public int traverse(TreeNode node) {
        depth++; 
        if(node == null) return depth--;

        ArrayList<Integer> levelList = levels.get(depth); //현재 트리 레벨의 노드 목록을 가져온다.
        if(levelList == null) levelList = new ArrayList<>();

        levelList.add(node.val); //현재 레벨의 노드 목록에 추가한다.
        levels.put(depth, levelList);

        traverse(node.left);
        traverse(node.right);

        return depth--;
    }
}

File: 0~100/M-98-Validate Binary Search Tree/JeongYe.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {Number} minValue 루트의 최소값
 * @param {Number} maxValue 루트의 최대값
 * @return {boolean}
 */
var isValidBST = function (root, minValue, maxValue) {
  if (!root) return true;

  // minValue <= root <= maxValue를 충족하지 않는 경우
  if (root.val <= minValue || maxValue <= root.val) return false;

  /**
   * root기준으로 왼쪽노드 < root , root < 오른쪽노드 조건이 성립하는지 재귀로 확인한다.
   * 왼쪽노드 탐색 : 왼쪽노드는 root 보다 항상 작기 때문에, 최대값이 root로 갱신된다.
   * 오른쪽노드 탐색 : 오른쪽노드는 root 보다 항상 크기 때문에, 최소값이 root로 갱신된다.
   */
  return isValidBST(root.left, minValue, root.val) && isValidBST(root.right, root.val, maxValue);
};
